server/src/main/java/io/crate/planner/optimizer/costs/PlanStats.java

```diff
@@ -56,6 +57,7 @@ public class PlanStats {
     // Memo can be null when there are no Group References in the Logical Plans involved
     @Nullable
     private final Memo memo;
+    private final IdentityHashMap<LogicalPlan, Stats> cache = new IdentityHashMap<>();
```

Comment:
LogicalPlan has no hashcode implementation therefore a ordinary map would also work, but this makes sure it will always be the the reference no matter what LogicalPlan implements in the future.

Reply:
Maybe it's a good moment to add some benchmarks to the optimizer anyway.